### DISCUSSION TITLE
v4-ovis-list

### DISCUSSION THREAD

**Discussion by narategithub** [narategithub/ldms#11] [General]
@tom95858

## Set format
From this:
https://github.com/narategithub/ovis/blob/2729bab9fb554fd1d72073a85dace2c8097b4e79/ldms/src/core/ldms.c#L1273-L1280

It looks like the set layout is as follows:
```python
[ meta_data | data0 | heap0 | data1 | heap1 | ... | data N | heap N ]
```

❗❓ **I want to verify that this is true.**

## how heap behave with set array
In the case of `set_array_card > 1`, when the sampler calls `ldms_transaction_begin()`, ldms assign `set->data` to the next set data entry. Then, the sampler usually set all metric values. 

https://github.com/narategithub/ovis/blob/2729bab9fb554fd1d72073a85dace2c8097b4e79/ldms/src/core/ldms.c#L2665-L2689

`set->heap` did not point to the new data entry though. This looks like a bug, assuming that the set has the memory layout above. Should `set->heap` points to the next heap memory when `set->data` points to the next data entry?

❗❓ If `set->heap` shall be set to the next heap memory the same way `set->data` does, shall we:
- copy the pervious heap data into the new heap memory? So that the application doesn't have to fill everything from scratch, or
- leave the new heap as-is and let the application fill everything from scratch?

**Comment by tom95858** [comment]
> set->heap did not point to the new data entry though. This looks like a bug, 

yes, no matter what, I think it's a bug

> ... assuming that the set has the memory layout above. Should set->heap points to the next heap memory when set->data points to the next data entry?

I think it has to, but I don't think it's implemented. I think when I prototyped this, I wasn't thinking about set buffers.

>
>❗❓ If set->heap shall be set to the next heap memory the same way set->data does, shall we:
>
>    copy the pervious heap data into the new heap memory? So that the application doesn't have to fill everything from scratch, or

I think the app has to build the lists items "from scratch" regardless, otherwise the whole purpose of the list becomes mute. The point is that a list metric may have more or fewer entries at each update without requiring that the set be deleted and reconstituted (like an array).

>    leave the new heap as-is and let the application fill everything from scratch?

I think it has to be both. But we might want to have an API that says ... "no change" in which case, it copies the old heap. But thinking about the current model, the sampler always sets every value in the metric set.

I think we need to think this through...

**Comment by narategithub** [comment]
@tom95858 

An update. https://github.com/narategithub/ovis/tree/v4-ovis-list is now v4.3.3-compatible. I also added a small python test scripts that showcase the list capability in https://github.com/narategithub/ovis/blob/v4-ovis-list/ldms/python/test/ldms_list/samp.py and https://github.com/narategithub/ovis/blob/v4-ovis-list/ldms/python/test/ldms_list/agg.py (with common routine: https://github.com/narategithub/ovis/blob/v4-ovis-list/ldms/python/test/ldms_list/common.py). You need to run `samp.py` first in a terminal, then `agg.py` in another terminal. They're currently being interactive so that I can manually update the data in samp, and call `s0.update()` in `agg.py` to update and `print_set(s0)` to see the change and `verify_set(s0)` to verify the data.

Here's the summary of what I modified on top of your prototype to make it compatible with v4.3.3:
- `ldms_set_hdr.data_sz` is data size + heap size. And, turned out we don't need heap_sz in the `ldms_set_hdr`.
- remove `heap_len` from the `ldms_rendezvous_lookup_param`. The downstream process (e.g. agg or ldms_ls) will know about the heap size when lookup completed. In the mean time, the newly-recorded set will be `inconsistent` and the application should not access the heap  data.
- Currently, when `ldms_transaction_begin()` is called, the heap and all list metrics are reset.

Things to consider:
- Shall I add `ldms_transaction_begin_preserved()` that will preserve the existing list data? (by copying, except when the `card` is 1, we don't have to do anything).
- In any case, the prototype is still missing the implementation of `ldms_list_del()`. Shall I implement it now, or it is something for the future?

Thanks,
Narate
